src/filter.rs

```diff
@@ -185,47 +264,67 @@ impl<'a, 'b> RayTracing<'a, 'b> {
             );
         }
 
-        if output.len() != buffer_dims {
-            return Err(FilterError::InvalidImageDimensions);
+        if output.len() != self.img_dims.2 {
+            return Err(Error::InvalidImageDimensions);
         }
         unsafe {
-            oidnSetSharedFilterImage(
+            oidnRetainDevice(self.device.0);
+        }
+        let output_buffer = unsafe {
+            oidnNewBuffer(self.device.0, byte_size)
+        };
+        unsafe {
+            oidnSetFilterImage(
                 self.handle,
                 b"output\0" as *const _ as _,
-                output.as_mut_ptr() as *mut _,
+                output_buffer,
                 OIDNFormat_OIDN_FORMAT_FLOAT3,
                 self.img_dims.0 as _,
                 self.img_dims.1 as _,
                 0,
                 0,
                 0,
             );
-        }
-
-        unsafe {
-            oidnSetFilter1b(self.handle, b"hdr\0" as *const _ as _, self.hdr);
-            oidnSetFilter1f(
+            oidnSetFilterBool(self.handle, b"hdr\0" as *const _ as _, self.hdr);
+            oidnSetFilterFloat(
                 self.handle,
                 b"inputScale\0" as *const _ as _,
                 self.input_scale,
             );
-            oidnSetFilter1b(self.handle, b"srgb\0" as *const _ as _, self.srgb);
-            oidnSetFilter1b(self.handle, b"clean_aux\0" as *const _ as _, self.clean_aux);
+            oidnSetFilterBool(self.handle, b"srgb\0" as *const _ as _, self.srgb);
+            oidnSetFilterBool(self.handle, b"clean_aux\0" as *const _ as _, self.clean_aux);
+
+            oidnSetFilterInt(self.handle, b"quality\0" as *const _ as _, self.filter_quality);
 
             oidnCommitFilter(self.handle);
             oidnExecuteFilter(self.handle);
+
+            oidnReadBuffer(output_buffer, 0, byte_size, output.as_mut_ptr() as *mut _);
+
+            oidnReleaseBuffer(color_buffer);
+            oidnReleaseBuffer(output_buffer);
         }
         Ok(())
     }
 }
 
-impl<'a, 'b> Drop for RayTracing<'a, 'b> {
+impl<'a> Drop for RayTracing<'a> {
     fn drop(&mut self) {
         unsafe {
             oidnReleaseFilter(self.handle);
             oidnReleaseDevice(self.device.0);
+            if let Some(norm) = self.normal {
+                oidnReleaseBuffer(
```

Comment:
I think what is missing here is releasing the additional device references we made when creating these additional OIDN buffers in `albedo_normal` / `albedo` . However since the `RayTracing` filter already keeps a device reference to keep the device alive, I don't think we need to make extra device retain calls for the buffers?

Reply:
I presume with the other change this is not necessary now?

Reply:
Correct, since we don't have the extra retains we don't need to do additional release's here